bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerFragment.java

```diff
@@ -53,8 +53,13 @@ public LedgerFragment(LedgerHandle lh,
         this.schedule = lh.getDistributionSchedule();
         SortedMap<Long, ? extends List<BookieId>> ensembles = lh
                 .getLedgerMetadata().getAllEnsembles();
+        // Check if the ledger fragment is closed has two conditions
+        // 1. The ledger is closed
+        // 2. This fragment is not the last fragment and this ledger's lastAddConfirm >= ensembles.lastKey() - 1.
+        //    This case happens when the ledger's last ensemble is empty
         this.isLedgerClosed = lh.getLedgerMetadata().isClosed()
-                || !ensemble.equals(ensembles.get(ensembles.lastKey()));
+                || (!ensemble.equals(ensembles.get(ensembles.lastKey()))
+            && lh.getLastAddConfirmed() >= ensembles.lastKey() - 1);
```

Comment:
Here should be `lh.getLastAddConfirmed() >= ensembles.lastKey() - 2`.
The last entry in the bookie may be entry 9, the getLastAddConfirmed is 9-1 = 8. The last ensemble key maybe 10. 8 >= 10 - 2.

Reply:
Maybe not. Take the following case for example:
0: [bk1, bk2, bk3]
10: [bk1, bk2, bk4]
LAC = 100
Ledger is OPEN

For the first ensemble: isLedgerClosed = (false || (true && 100 >= 9)) => true
For the last ensemble: isLedgerClosed = (false || (false && 100 >= 9)) => false
The result is expected.

Reply:
Yes, you are right.